### ISSUE TITLE
Backtesting & Prediction System for 1001 Movies List

### ISSUE THREAD

**Issue by holden** [razrfly/cinegraph#287]
# Backtesting & Prediction System for 1001 Movies List

## Overview
Now that we have all data sources integrated (ratings, awards, PQS, cultural metrics), we need to build a comprehensive backtesting and prediction system that can:
1. **Backtest**: Replicate the existing 1001 Movies list with >90% accuracy using our weighted metrics
2. **Predict**: Forecast which movies from recent years (2020-2025) are most likely to be added to future editions

## Current Status
- ✅ All data sources integrated (TMDb, IMDb, OMDb, festivals, canonical lists)
- ✅ PQS (Person Quality Score) system implemented
- ✅ Cultural Relevance Index (CRI) framework designed (#260)
- ✅ Initial metrics audit complete (#229) - 50% data coverage
- ✅ Normalization and weighting system in place
- ❌ Need: Decade-based analysis and constraints
- ❌ Need: Backtesting optimization algorithm
- ❌ Need: Prediction engine for future entries
- ❌ Need: Visual interface for analysis and tweaking

### 1. Decade-Based Analysis
The 1001 Movies list has inherent decade-based distribution patterns:
- Calculate average movies per decade in the existing list
- Implement decade-based quotas/constraints in our algorithm
- Account for recency bias in newer decades
- Handle the "too recent to judge" factor for movies <5 years old

#### Phase 1: Replication Algorithm
- Use only our 6 metric categories (no cheating by using list membership as a factor):
  1. Critical Reception (IMDb, TMDb, RT, Metacritic)
  2. Commercial Success (box office, adjusted for inflation)
  3. Cultural Impact (canonical list memberships, festival wins)
  4. Artistic Merit (PQS scores, director/cast quality)
  5. Historical Significance (first-of-its-kind, influence metrics)
  6. Audience Engagement (vote counts, popularity scores)

#### Phase 2: Weight Optimization
**Approach Options:**
1. **Genetic Algorithm**: Evolve weight combinations over generations
2. **Gradient Descent**: Optimize weights to minimize prediction error
3. **Grid Search**: Systematic exploration of weight space
4. **Bayesian Optimization**: Smart sampling of weight combinations
5. **Ensemble Methods**: Combine multiple approaches

**Proposed Solution**: Start with genetic algorithm for global optimization, then fine-tune with gradient descent

### 3. Prediction Engine
For movies 2020-2025 not yet in the list:
- Calculate probability scores (0-100%) for inclusion
- Rank by likelihood within decade constraints
- Account for:
  - Time decay (newer movies need higher scores)
  - Genre representation gaps
  - Geographic/language diversity
  - Director/actor representation

#### Dashboard Components:

**A. Backtesting View**
```
┌─────────────────────────────────────────┐
│         Backtesting Dashboard           │
├─────────────────────────────────────────┤
│ Overall Accuracy: [====92%====]         │
│                                         │
│ By Decade:                             │
│ 1920s: [===95%===] (12/13 correct)     │
│ 1930s: [===88%===] (28/32 correct)     │
│ 1940s: [===91%===] (41/45 correct)     │
│ ...                                     │
│ 2010s: [===87%===] (78/90 correct)     │
└─────────────────────────────────────────┘
```

**B. Weight Tuning Interface**
- Interactive sliders for each metric category
- Real-time accuracy recalculation
- Heatmap showing weight sensitivity
- Save/load weight presets
- A/B comparison of weight sets

**C. Prediction View**
```
┌─────────────────────────────────────────┐
│      2020s Predictions (Top 20)        │
├─────────────────────────────────────────┤
│ 1. Oppenheimer (2023) - 94% chance     │
│ 2. Everything Everywhere... - 91%       │
│ 3. The Power of the Dog - 88%          │
│ 4. Dune (2021) - 85%                   │
│ ...                                     │
│                                         │
│ Decade Quota: ~15-20 films expected    │
└─────────────────────────────────────────┘
```

**D. Analysis Tools**
- False positive/negative analysis
- Feature importance visualization
- Confusion matrix by decade
- ROC curves and precision-recall curves
- Distribution plots of scores

### Database Schema Additions
```sql
-- Backtesting results
CREATE TABLE backtest_runs (
  id UUID PRIMARY KEY,
  weights JSONB NOT NULL, -- Weight configuration used
  accuracy_overall FLOAT,
  accuracy_by_decade JSONB,
  false_positives INTEGER,
  false_negatives INTEGER,
  created_at TIMESTAMP
);

-- Prediction scores
CREATE TABLE prediction_scores (
  movie_id UUID REFERENCES movies(id),
  backtest_run_id UUID REFERENCES backtest_runs(id),
  probability FLOAT, -- 0-1 probability of inclusion
  rank_in_decade INTEGER,
  contributing_factors JSONB,
  created_at TIMESTAMP
);

-- Decade constraints
CREATE TABLE decade_targets (
  decade INTEGER PRIMARY KEY,
  target_count INTEGER,
  actual_count INTEGER,
  notes TEXT
);
```

### Algorithm Components

1. **Weight Optimizer Module** (`lib/cinegraph/backtesting/optimizer.ex`)
   - Genetic algorithm implementation
   - Fitness function: minimize (false_positives + false_negatives)
   - Constraints: decade quotas, weight bounds

2. **Prediction Engine** (`lib/cinegraph/predictions/engine.ex`)
   - Score calculation pipeline
   - Decade-aware ranking
   - Confidence interval estimation

3. **Backtesting Service** (`lib/cinegraph/backtesting/service.ex`)
   - Run backtests with different weight configurations
   - Generate accuracy metrics
   - Store results for comparison

### LiveView Interface

```elixir
# lib/cinegraph_web/live/backtest_live.ex
- Main dashboard with tabs for backtesting/prediction
- Real-time weight adjustment with WebSocket updates
- D3.js or Chart.js visualizations

# lib/cinegraph_web/live/prediction_live.ex  
- Sortable/filterable prediction results
- Detailed breakdown of scoring factors
- Export functionality for reports
```

### Phase 1: Data Preparation (Week 1)
- [ ] Calculate decade distributions in existing list
- [ ] Fill remaining data gaps (reach 80%+ coverage)
- [ ] Create training/validation split (80/20)
- [ ] Implement decade constraint logic

### Phase 2: Backtesting Engine (Week 2)
- [ ] Implement genetic algorithm optimizer
- [ ] Create fitness evaluation function
- [ ] Build accuracy measurement system
- [ ] Develop weight persistence layer

### Phase 3: Prediction System (Week 3)
- [ ] Build prediction scoring pipeline
- [ ] Implement decade-aware ranking
- [ ] Create confidence scoring
- [ ] Handle edge cases (recent films, missing data)

### Phase 4: Interface Development (Week 4)
- [ ] Create LiveView dashboard
- [ ] Implement weight tuning interface
- [ ] Build visualization components
- [ ] Add export/reporting features

### Phase 5: Optimization & Testing (Week 5)
- [ ] Fine-tune algorithm parameters
- [ ] Cross-validation testing
- [ ] Performance optimization
- [ ] Documentation and testing

## Success Metrics
- **Primary**: Achieve >90% backtesting accuracy
- **Secondary**: <5% false positive rate per decade
- **Tertiary**: Meaningful predictions with clear rationale

## Open Questions
1. Should we use ensemble methods combining multiple algorithms?
2. How do we handle films with incomplete data (imputation strategies)?
3. Should we incorporate temporal factors (e.g., films gain cultural relevance over time)?
4. Do we need different weight sets for different decades?
5. How do we validate predictions before the next edition comes out?

## ML Library Options
- **Nx/Axon**: Native Elixir ML (preferred for integration)
- **Python Interop**: Use scikit-learn via Ports/NIFs if needed
- **Custom Implementation**: Pure Elixir for genetic algorithms (simpler, more control)

## Related Issues
- #260 - Cultural Relevance Index (CRI) design
- #229 - Initial data audit
- #3 - Original 1001 Movies tracking
- #4 - Prediction model requirements

## Next Steps
1. Review and refine requirements
2. Choose ML approach (start with genetic algorithm)
3. Create database migrations for new tables
4. Begin Phase 1 implementation

---

**Labels**: enhancement, machine-learning, analytics, ui/ux, high-priority
**Assignees**: @razrfly
**Milestone**: 1001 Movies Prediction System

**Comment by holden** [comment]
DUPLICATE of #320 — backtesting + prediction engine fully shipped.
